File: node/vuo.app/vuo.app.launch.c
/**
 * @file
 * vuo.app.launch node implementation.
 *
 * @copyright Copyright © 2012–2023 Kosada Incorporated.
 * This code may be modified and distributed under the terms of the MIT License.
 * For more information, see https://vuo.org/license.
 */

#include "VuoAppLaunch.h"

VuoModuleMetadata({
					 "title" : "Launch App",
					 "keywords" : [ "execute", "run", "start", "open", "application" ],
					 "version" : "1.2.0",
					 "dependencies" : [
						 "VuoAppLaunch"
					 ],
					 "node": {
						 "exampleCompositions" : [ "LaunchCalculator.vuo" ]
					 }
				 });

void nodeEvent
(
		VuoInputData(VuoText, {"name":"URL", "default":"Calculator.app"}) url,
		VuoInputEvent({"eventBlocking":"none","data":"url"}) urlEvent,
		VuoInputData(VuoBoolean, {"default":true}) activate
)
{
	if (urlEvent)
		VuoAppLaunch_launch(url, activate);
}

File: node/vuo.app/VuoAppLaunch.m
/**
 * @file
 * VuoAppLaunch implementation.
 *
 * @copyright Copyright © 2012–2021 Kosada Incorporated.
 * This code may be modified and distributed under the terms of the MIT License.
 * For more information, see https://vuo.org/license.
 */

#include "VuoAppLaunch.h"

#include "VuoMacOSSDKWorkaround.h"
#include <AppKit/AppKit.h>

#include "module.h"

#ifdef VUO_COMPILER
VuoModuleMetadata({
					"title" : "VuoAppLaunch",
					"dependencies" : [
						"VuoUrl",
						"AppKit.framework"
					]
				 });
#endif

/**
 * Launches the `.app` at `rawUrl`, or brings it to the foreground if it's already running.
 */
void VuoAppLaunch_launch(VuoText rawUrl, bool activate)
{
	if (VuoText_isEmpty(rawUrl))
		return;

	VuoUrl normalizedUrl = VuoUrl_normalize(rawUrl, VuoUrlNormalize_forLaunching);
	NSString *urlAsString = [[NSString alloc] initWithUTF8String:normalizedUrl];
	NSURL *url = [[NSURL alloc] initWithString:urlAsString];
	[urlAsString release];

	NSError *error = nil;
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
	// The replacement, `-[NSWorkspace openApplicationAtURL]`, isn't available until macOS 11.
	NSRunningApplication *app = [[NSWorkspace sharedWorkspace] launchApplicationAtURL:url
																			  options:(activate ? 0 : NSWorkspaceLaunchWithoutActivation)
																		configuration:@{}
																				error:&error];
#pragma clang diagnostic pop
	if (! app)
		VUserLog("Couldn't launch '%s': %s — %s", normalizedUrl,
				 [[error localizedDescription] UTF8String],
				 [[[[error userInfo] objectForKey:NSUnderlyingErrorKey] localizedDescription] UTF8String]);

	VuoRetain(normalizedUrl);
	VuoRelease(normalizedUrl);
	[url release];
}
